miniapps/adjoint/advection_diffusion.cpp

```diff
@@ -0,0 +1,488 @@
+//                [  int_t int_x (v * du / dx) dx dt     ]
+
+// This version uses MPI for user routines.
+// Execute with Number of Processors = N,  with 1 <= N <= MX.
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+#include <algorithm>
+
+#ifndef MFEM_USE_SUNDIALS
+#error This example requires that MFEM is built with MFEM_USE_SUNDIALS=YES
+#endif
+
+using namespace std;
+using namespace mfem;
+
+// Implement the adjoint rate equations in AdvDiffSUNDIALS
+class AdvDiffSUNDIALS : public TimeDependentAdjointOperator
+{
+public:
+   AdvDiffSUNDIALS(int ydot_dim, int ybdot_dim, Vector p,
+                   ParFiniteElementSpace *fes, Array<int> & ess_tdof) :
+      TimeDependentAdjointOperator(ydot_dim, ybdot_dim),
+      p_(p),
+      M(NULL), K(NULL), K_adj(NULL),
+      Mf(NULL), MK(NULL),
+      m(NULL), k(NULL),
+      pfes(fes),
+      M_solver(fes->GetComm()),
+      ess_tdof_list(ess_tdof)
+   {
+      int skip_zeros = 0;
+      ParMesh * pmesh = pfes->GetParMesh();
+
+      cout << "Essential tdofs: " << endl;
+      ess_tdof_list.Print();
+
+      m = new ParBilinearForm(pfes);
+      m->AddDomainIntegrator(new MassIntegrator());
+      m->Assemble(skip_zeros);
+      m->Finalize(skip_zeros);
+
+      k = new ParBilinearForm(pfes);
+      k->AddDomainIntegrator(
+	   new DiffusionIntegrator(
+		  *(new ConstantCoefficient(-p_[0]))));
+      Vector p2(fes->GetParMesh()->SpaceDimension());
+      p2 = p_[1];
+      k->AddDomainIntegrator(
+	   new ConvectionIntegrator(
+		  *(new VectorConstantCoefficient(p2))));
+      k->Assemble(skip_zeros);
+      k->Finalize(skip_zeros);
+
+      k1 = new ParBilinearForm(pfes);
+      k1->AddDomainIntegrator(
+	   new DiffusionIntegrator(
+		  *(new ConstantCoefficient(p_[0]))));
+      k1->AddDomainIntegrator(
+	   new ConvectionIntegrator(
+		  *(new VectorConstantCoefficient(p2))));
+      k1->Assemble(skip_zeros);
+      k1->Finalize(skip_zeros);
+
+      M = m->ParallelAssemble();
+      M->EliminateRowsCols(ess_tdof_list);
+
+      K = k->ParallelAssemble();
+      K->EliminateRowsCols(ess_tdof_list);
+
+      K_adj = k1->ParallelAssemble();
+      K_adj->EliminateRowsCols(ess_tdof_list);
+
+      MK = ParMult(M, K);
+
+      M_prec.SetType(HypreSmoother::Jacobi);
+      M_solver.SetPreconditioner(M_prec);
+      M_solver.SetOperator(*M);
+
+      M_solver.SetRelTol(1e-14);
+      M_solver.SetAbsTol(0.0);
+      M_solver.SetMaxIter(1000);
+      M_solver.SetPrintLevel(0);
+
+   }
+
+   virtual void Mult(const Vector &x, Vector &y) const;
+  
+   virtual void AdjointRateMult(const Vector &y, Vector &yB,
+				Vector &yBdot) const;
+
+   virtual int SUNImplicitSetup(const Vector &y,
+                                const Vector &fy, int jok, int *jcur,
+				double gamma);
+  
+   virtual int SUNImplicitSolve(const Vector &b, Vector &x, double tol);
+
+  virtual void QuadratureSensitivityMult(const Vector &y, const Vector &yB,
```

Comment:
The spacing here seems strange. Does this pass the style check?

Reply:
It does pass astyle (Artistic Style Version 2.05.1). Since it's a miniapp do I need to do anything special?